pkg/upload/dsym.go

```diff
@@ -0,0 +1,165 @@
+package upload
+
+import (
+	"path/filepath"
+
+	"github.com/bugsnag/bugsnag-cli/pkg/ios"
+	"github.com/bugsnag/bugsnag-cli/pkg/log"
+	"github.com/bugsnag/bugsnag-cli/pkg/server"
+	"github.com/bugsnag/bugsnag-cli/pkg/utils"
+)
+
+type Dsym struct {
+	VersionName string      `help:"The version of the application."`
+	Scheme      string      `help:"The name of the scheme to use when building the application."`
+	Dev         bool        `help:"Indicates whether the application is a debug or release build"`
+	DsymPath    string      `help:"Path to the dSYM" type:"path"`
+	Plist       string      `help:"Path to the Info.plist file" type:"path"`
+	ProjectRoot string      `help:"path to remove from the beginning of the filenames in the mapping file" type:"path"`
+	Path        utils.Paths `arg:"" name:"path" help:"Path to directory or file to upload" type:"path" default:"."`
+}
+
+func ProcessDsym(
+	apiKey string,
+	versionName string,
+	scheme string,
+	dev bool,
+	dsymPath string,
+	plistPath string,
+	projectRoot string,
+	paths []string,
+	endpoint string,
+	timeout int,
+	retries int,
+	overwrite bool,
+	dryRun bool,
+) error {
+
+	var buildSettings *ios.XcodeBuildSettings
+	var dsyms *[]*ios.DwarfInfo
+	var plistData *ios.PlistData
+	var uploadOptions map[string]string
+
+	for _, path := range paths {
+		uploadInfo, err := ios.ProcessPathValue(path, projectRoot)
+		if err != nil {
+			return err
+		}
+
+		// If dsymPath is not set explicitly, use uploadInfo to set it (if available)
+		// uploadInfo.DsymPath is set when <path> is recognised as a dsym path
+		if dsymPath == "" && uploadInfo.DsymPath != "" {
+			dsymPath = uploadInfo.DsymPath
+		}
+
+		// If projectRoot is not set explicitly, use uploadInfo to set it
+		if projectRoot == "" {
+			projectRoot = uploadInfo.ProjectRoot
+		}
```

Comment:
It feels to me that this section is a little over-complicated with the intermediate struct – is there any reason not to have separate functions to set the project root (if not provided) and dSYM path (if not provided); or use multiple return values from `ProcessPathValue`? (However to me, it feels like `ProcessPathValue` is trying to do too much and making it harder to read in doing so.)

Reply:
I'm in agreement that `ProcessPathValue` is a little "busy" in it's current form. I'll see what I can do to split things out a bit better and maybe pull out some of these empty checks into an intermediate validation function or something perhaps?

Reply:
IMO there doesn't seem much point in the `DsymUploadInfo` struct now that we've established that project root isn't going to be used for any other input and we're not defaulting it to cwd. My suggestion would be to have 2 functions and handle the override, validation and default values for `projectRoot` and `path` separately.